### PR TITLE
Get rid of user-specified refs

### PR DESCRIPTION
**Pull request by jfirebaugh**
WIP for #1393. Depends on https://github.com/mapbox/mapbox-gl-style-spec/pull/553. Further development work waiting on https://github.com/mapbox/mapbox-gl-style-spec/issues/554.

### CODE REVIEW

**Comment by tmcw** [comment]
🎉  so psyched to see this happening, thx for working on it @jfirebaugh!

**Comment by lucaswoj** [comment]
This is great! I love how simple `groupByLayout` is.

**Comment by jfirebaugh** [comment]
Rebased after #3487 and this is now passing.

File: js/data/bucket.js [hunk 1]
```diff
@@ -57,10 +57,19 @@ class Bucket {
...
+     */
     destroy() {
-        this.buffers.destroy();
-    }
-}
+        if (this.buffers) {
+            this.buffers.destroy();
+            this.buffers = null;
+        }
+    }}
```
Comment (mourner) [inline js/data/bucket.js]: weird indenting here

**Comment by jfirebaugh** [comment]
benchmark | master 74077d7 | deref 3d1ff7f
--- | --- | ---
**map-load** | 143 ms  | 110 ms 
**style-load** | 127 ms  | 127 ms 
**buffer** | 1,004 ms  | 987 ms 
**fps** | 60 fps  | 60 fps 
**frame-duration** | 4.2 ms, 0% > 16ms  | 4.2 ms, 0% > 16ms 
**query-point** | 0.88 ms  | 0.90 ms 
**query-box** | 72.87 ms  | 74.32 ms 
**geojson-setdata-small** | 7 ms  | 6 ms 
**geojson-setdata-large** | 230 ms  | 265 ms

**Pull request merged**